src/github.com/getlantern/netx/copy.go

```diff
@@ -0,0 +1,67 @@
+package netx
+
+import (
+	"io"
+	"net"
+	"sync/atomic"
+	"time"
+)
+
+var (
+	copyTimeout = 1 * time.Second
+)
+
+// BidiCopy copies between in and out in both directions using the specified
+// buffers, returning the errors from copying to out and copying to in.
+func BidiCopy(out net.Conn, in net.Conn, bufOut []byte, bufIn []byte) (outErr error, inErr error) {
+	stop := uint32(0)
+	outErrCh := make(chan error, 1)
+	inErrCh := make(chan error, 1)
+	go doCopy(out, in, bufIn, outErrCh, &stop)
+	go doCopy(in, out, bufOut, inErrCh, &stop)
+	return <-outErrCh, <-inErrCh
+}
+
+// doCopy is based on io.copyBuffer
+func doCopy(dst net.Conn, src net.Conn, buf []byte, errCh chan error, stop *uint32) {
+	var err error
+	defer func() {
+		atomic.StoreUint32(stop, 1)
+		errCh <- err
+	}()
+
+	for {
+		if atomic.LoadUint32(stop) == 1 {
```

Comment:
If we continue to loop here if no error, whether it's timeout or not, has occurred in previous iteration, the flaw I mentioned is fixed.